File: app_mobile_legacy/src/main/java/com/kee0kai/thekey/engine/FindStorageEngine.java
package com.kee0kai.thekey.engine;

import com.kee0kai.thekey.model.Storage;

public class FindStorageEngine {

    static {
        System.loadLibrary("crypt-storage-lib");
    }

    private static FileFindListener findListener = null;


    public native void findStorage(String sourceDir);


    public void setFindListener(FileFindListener findListener) {
        FindStorageEngine.findListener = findListener;
    }

    // from Jni
    private void onStorageFounded(Storage storage) {
        if (findListener != null)
            findListener.onStorageFound(storage);
    }

    public interface FileFindListener {
        void onStorageFound(Storage storage);
    }

}

File: app_mobile/src/main/kotlin/com/github/klee0kai/thekey/app/engine/FindStorageEngine.kt
package com.github.klee0kai.thekey.app.engine

import com.github.klee0kai.brooklyn.Brooklyn
import com.github.klee0kai.brooklyn.JniMirror
import com.github.klee0kai.thekey.app.di.DI
import com.github.klee0kai.thekey.app.model.Storage
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.launch

@JniMirror
open class FindStorageEngine {

    init {
        Brooklyn.loadLibrary("crypt-storage-lib")
    }

    open external fun findStorages(folder: String, listener: FindStorageListener)

}

fun FindStorageEngine.findStoragesFlow(folder: String): Flow<Storage> = callbackFlow {
    launch(DI.ioDispatcher()) {
        findStorages(folder, object : FindStorageListener() {
            override fun onStorageFound(storage: Storage) {
                launch { send(storage) }
            }
        })

        channel.close()
    }

    awaitClose()
}